Fix SELECT FOR XML/JSON behavior

Sairakan opened the pull request:
### Description

SELECT FOR XML/JSON were both originally implemented by converting the select query into a function call that then called SPI_execute() against the original SELECT statement. While that implementation works for simple SELECT statements, it fails to execute properly against queries where the SELECT statement is a subquery which references CTEs or correlated aliases outside of the scope of the sub-SELECT. This is due to the fact that the SELECT statement is passed as a string argument to the internal function call for SPI. Thus, any external references are lost when SPI_execute() is called.

We resolve this issue by changing the implementation of SELECT FOR XML/JSON to use aggregate functions. This way, there is no need to store any query strings, and the executor will fully execute the sub-SELECT prior to passing the result rows to the aggregate function, allowing us to retain the context of any outer CTEs or correlations.

Signed-off-by: Jason Teng <jasonten@amazon.com>

Matching engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/65
 
### Issues Resolved

BABEL-1876

### Testing

- [x] Regression testing against older tests
- [x] Test cases with correlated CTEs/subqueries

Performance testing results:
|# of rows | SQL Server | Old Implementation | New Implemenation
--- | --- | --- | ---
|10,000 | 10 ms | 180 ms | 176 ms|
|100,000 | 435 ms | 1816 ms | 1774 ms|
|1,000,000 | 8047 ms | 18568 ms | 18066 ms|

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -102,6 +102,7 @@ Test-Computed-Columns
 BABEL-1189
 BABEL-1243
 BABEL-1493
+BABEL-1876
```
Latest is only for self MVU. This change has minor upgrade implications as well. Add in 14.5 schedule as well.

Sairakan replied:
Added

2jungkook commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -684,7 +684,6 @@ Function sys.tsql_get_constraintdef(oid)
 Function sys.tsql_get_functiondef(oid)
 Function sys.tsql_get_returntypmodvalue(oid)
 Function sys.tsql_query_to_xml(text,integer,text,boolean,text)
```
Could you add a test for this function as well?

2jungkook commented on test/JDBC/expected/BABEL-1876-vu-verify.out:
```diff
@@ -0,0 +1,14 @@
+int#!#ntext
+~~END~~
+
+select * from babel_1876_vu_v1
```
Could you also check the cases where the correlated FOR XML subquery is in `where` or `from` clause?

Commit:
Redesign SELECT FOR XML/JSON to use aggregate function instead of SPI call

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Update upgrade script, fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Add tests to upgrade schedule files

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
fix compiler errors

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Formatting and comment updates, adding method stubs for major version upgrade compatibility

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
remove deleted test from upgrade schedule file

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Change name of alias to "rows"

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Add tsql_query_to_json_text stub for MVU compatibility

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix upgrade script typos in deprecation commands

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
fix typos

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
fix expected_dependency test

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
move internal function call tests to separate mvu test file

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Remove unnecessary function stub

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Revert "Remove unnecessary function stub"

This reverts commit df9ad0d0430a094049f4c5ce5d5ed450a9972829.

Commit:
Fix upgrade tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix expected test output

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Change "rows" to an md5 hash to avoid collision with user-defined column aliases

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
update code comments, move binary test to separate test files

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Update forxml tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Sairakan committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-1876

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -138,8 +138,10 @@ BABEL-3646
 BABEL-937
-BABEL-3407
 forjson
+forjson-datatypes
```
this should move to 14.6 schedule to test Narwal to Ocelot

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
fix merge conflict properly

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
update 14.6 upgrade schedule file

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
move direct function calls to separate test

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
try removing test from 14_5 schedule file

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Add forxml test back to 14_5 schedule file and try fixing crash by returning the correct type for the deprecated functions

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
fix compiler error

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
update deprecated functions to just return NULL

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Add back in full implementation of older versions of for xml/json functions

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
update tests to keep old behavior for old versions and separately test new features

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1,22 +1,123 @@
+    root_name text
+) RETURNS TEXT
+AS 'babelfishpg_tsql', 'tsql_query_to_xml_sfunc'
+LANGUAGE C COST 100;
```
Just wondering where did `COST 100` come from?

Sairakan replied:
This is what was used in the original function for `for xml` so I kept it here. I don't think it is actually necessary though

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1,22 +1,123 @@
+END IF;
+END;
+$$
+LANGUAGE PLPGSQL STRICT;
```
1. Do we have a test for `STRICT` keyword?
2. Can we use `IMMUTABLE` here?

Sairakan replied:
Any tests that return 0 rows will trigger the STRICT keyword. I believe there are a few such test cases.
It's probably okay to use IMMUTABLE, let me think about it a bit more.

Sairakan replied:
I've updated the sfuncs to STABLE and ffuncs to IMMUTABLE

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1,22 +1,123 @@
+AS 'babelfishpg_tsql', 'tsql_query_to_xml_sfunc'
+LANGUAGE C COST 100;
+
+CREATE OR REPLACE FUNCTION sys.tsql_query_to_xml_ffunc(
```
`tsql_query_to_xml_ffunc` and `tsql_query_to_xml_text_ffunc` are almost duplicate. Can we use `return tsql_query_to_xml_text_ffunc(...)::XML` in this function body?

Sairakan replied:
Good point, that should work and reduce code duplication.

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1369,40 +1219,85 @@ returning_list_has_column_name(List *existing_colnames, char *current_colname)
+		func_name= list_make2(makeString("sys"), makeString("tsql_select_for_xml_agg"));
+	else
+		func_name= list_make2(makeString("sys"), makeString("tsql_select_for_xml_text_agg"));
+	func_args = list_make5(makeColumnRef(construct_unique_index_name("rows", "forjson"), NIL, -1, NULL),
```
1. `forjson` seems not correct
2. I understand that you are trying to make a unique column name, but using `construct_unique_index_name()` could give a wrong impression to both developers and users. `construct_unique_index_name()` is doing many additional things and may throw an error like "constructing unique index name failed".
3. I would suggest to create a separate function for this purpose.

Sairakan replied:
1. I can change the name to something else, but it does need to be the same for both `for json` and `for xml` since they share the same `TsqlForClauseSubselect` usage. I'll probably use something like `tsql_for`.
2. `construct-unique-index-name()` is overall a pretty simple function, it just does an md5 hash of the first argument and then appends the second argument to the end - the only case when the error would be raised is if the md5 itself fails, which is extremely unlikely and would basically only happen if the system hits OOM per the comments in the md5 function. I don't think it is worth creating a new function for this case, but I'll add some comments explaining why this is used in the first place.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -1361,155 +1361,119 @@ tsql_UpdateStmt: opt_with_clause UPDATE relation_expr_opt_alias
-						insertSelectOptions((SelectStmt *) $1, $2, NIL,
-											NULL, NULL,
-											yyscanner);
+						$$ = $1;
```
In which case, this line could happen?

Sairakan replied:
to be honest, I am not sure. This is that pattern that was previously used, which is why I kept it here, but based on the grammar rules there doesn't seem to be a way to actually match tsql_for_xml_clause while generating a null value. I'll try removing it and see what happens.

2jungkook commented on contrib/babelfishpg_tsql/src/tsql_for/forjson.c:
```diff
@@ -0,0 +1,222 @@
+	HeapTupleHeader td;
+	Oid				tupType;
+	int32			tupTypmod;
+	TupleDesc		tupdesc;
```
We may need `ReleaseTupleDesc()`.

Sairakan replied:
Good point, added

2jungkook commented on contrib/babelfishpg_tsql/src/tsql_for/forxml.c:
```diff
@@ -0,0 +1,329 @@
+		 * tsql_datatype_oid can be different from datatype_oid when there are datatypes in different namespaces
+		 * but with the same name. Examples: bigint, int, etc.
+		 */
+		if (tsql_datatype_oid == datatype_oid)
```
What about `bit` and `money` types? Do they behave differently than `for json`?

Sairakan replied:
Yeah, bit and money have a different behavior compared with `for json` - they are not coerced to boolean/numeric types and are left as-is with `for xml`

2jungkook commented on contrib/babelfishpg_tsql/src/tsql_for/forxml.c:
```diff
@@ -0,0 +1,329 @@
+		 * tsql_datatype_oid can be different from datatype_oid when there are datatypes in different namespaces
+		 * but with the same name. Examples: bigint, int, etc.
+		 */
+		if (tsql_datatype_oid == datatype_oid)
```
This check logic can be shared at least between `tsql_row_to_xml_path()` and `tsql_row_to_xml_raw()`.

Sairakan replied:
Extracted the shared code into a separate function.

2jungkook commented on contrib/babelfishpg_tsql/src/tsql_for/forxml_old.c:
```diff
@@ -45,6 +52,10 @@ tsql_query_to_xml(PG_FUNCTION_ARGS)
 
+	ereport(WARNING,
+			(errcode(ERRCODE_WARNING_DEPRECATED_FEATURE),
+			 errmsg("This version of FOR XML has been deprecated. We recommend recreating the view for this query.")));
```
There is no test for this. You can either
1. modify our JDBC test framework to test the warning message, or
2. use installcheck

2jungkook commented on contrib/babelfishpg_tsql/src/tsql_for/tsql_for.c:
```diff
@@ -0,0 +1,73 @@
+ * "2022-11-11 20:56:22.41" -> "2022-11-11T20:56:22.41" for datetime, datetime2 & smalldatetime
+ */
+void
+tsql_for_datetime_format(StringInfo format_output, char *outputstr)
```
Let's try to use `const char *` if we don't expect the input parameter be changed.

Sairakan replied:
Will do

2jungkook commented on test/JDBC/expected/forjson-after-14-6-vu-prepare.out:
```diff
@@ -0,0 +1,139 @@
+
+-- 14.6 (aka extension version 2.4.0) has a major change to how FOR JSON
```
1. BBF 2.4 corresponds to PG 14.7
2. Can we use `before-14-x` like other tests?

Sairakan replied:
fixed the comment to refer to 14.7, and renamed the tests to before-14-7 and normal tests

Commit:
Update aggregate state transition functions to pass-by-reference
in order to reduce unnecessary copying which was significantly
slowing down execution for large queries.

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Commit:
Remove direct function call tests (as they can no longer be executed) and fix root tag handling for xml

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
set functions to stable, update upgrade schedule files

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Add functions to sql_validation_framework expected output. Since
these are functions with type INTERNAL, there is no way to exercise
them manually to get rid of this output.

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
fix upgrade validation expected output

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
rename test files, address comments

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
fix function declaration

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
fix expected test output

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
extract shared code into separate function in forxml.c

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
rename variable

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
add code comments

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
add deprecation warning to forjson_old.c

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
allow FOR XML to accept arbitrarily long root tags

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
off-by-1 bug fix for root tag

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
set ffuncs to immutable, fix update script to match sys_functions.sql

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

Signed-off-by: Jason Teng <jasonten@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql:
```diff
@@ -157,6 +157,97 @@ CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 TINYINT)
+ALTER FUNCTION sys.tsql_query_to_xml_text(text, int, text, boolean, text) RENAME TO tsql_query_to_xml_text_deprecated_in_2_4_0;
+CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'tsql_query_to_xml_text_deprecated_in_2_4_0');
+
+ALTER FUNCTION sys.tsql_query_to_json_text(text, int, boolean, boolean, text) RENAME TO tsql_query_to_json_text_deprecated_in_2_4_0;
```
Please be careful when you make the same change for `3_X_DEV`. In the `3.0.0--3.1.0` upgrade script, `sys.tsql_query_to_json_text` may or may not exist because of two different upgrade paths:
1. 2.3 --> 3.0 --> 3.1: `sys.tsql_query_to_json_text` exists
2. 2.4 --> 3.0 --> 3.1: `sys.tsql_query_to_json_text` does not exists

2jungkook commented on test/JDBC/upgrade/14_5/schedule:
```diff
@@ -36,6 +36,7 @@ TestUDD
 TestUniqueIdentifier
 TestVarChar
 TestXML
+forxml-before-14-7
```
Let's rename this to `forxml-before-14-7-or-15-2` to have the same test names between 2_X_DEV and 3_X_DEV.

2jungkook commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -139,8 +139,9 @@ BABEL-3646
 BABEL-937
-BABEL-3407
-forjson
+forjson-before-14-7
```
Why do we have before-14-7 in here?

Sairakan replied:
hm, rethinking this, the name is not properly descriptive of what the tests are - these are supposed to be the regression tests that apply to both previous versions as well as any future versions. The normal forjson and forxml test suites then are only testing the new features added in 14.7. I think it would be best to go back to the original naming scheme, where these are the normal tests (aka forjson/forxml), and the other test files are named `after-14-7`

Sairakan replied:
after discussion, renaming this to forjson/forxml, and the other tests to forjson-subquery/forxml-subquery

Commit:
add new tests to latest schedule file

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
rename test files

Signed-off-by: Jason Teng <jasonten@amazon.com>

Sairakan committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-1876

Commit:
fix merge conflict properly

Signed-off-by: Jason Teng <jasonten@amazon.com>

2jungkook reviewed (dismissed):
Please add the engine PR link to the PR description

Commit:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-1876

2jungkook committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-1876
